transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.channel;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.ImmediateEventExecutor;
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.ArrayDeque;
+
+/**
+ * A FIFO queue of bytes where producers add bytes by repeatedly adding {@link ByteBuf} and consumers take bytes in
+ * arbitrary lengths. This allows producers to add lots of small buffers and the consumer to take all the bytes
+ * out in a single buffer. Conversely the producer may add larger buffers and the consumer could take the bytes in
+ * many small buffers.
+ *
+ * <p>Bytes are added and removed with promises. If the last byte of a buffer added with a promise is removed then
+ * that promise will complete when the promise passed to {@link #remove} completes.
+ *
+ * <p>This functionality is useful for aggregating or partitioning writes into fixed size buffers for framing protocols
+ * such as HTTP2.
+ */
+public final class CoalescingBufferQueue {
+
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(CoalescingBufferQueue.class);
+
+    private final ByteBufAllocator alloc;
+    private final ArrayDeque<Object> bufAndListenerPairs = new ArrayDeque<Object>();
+    private int readableBytes;
+
+    public CoalescingBufferQueue(ByteBufAllocator alloc) {
+        this.alloc = ObjectUtil.checkNotNull(alloc, "alloc");
+    }
+
+    /**
+     * Add a buffer to the end of the queue.
+     */
+    public void add(ByteBuf buf) {
+        ObjectUtil.checkNotNull(buf, "buf");
+        if (readableBytes > Integer.MAX_VALUE - buf.readableBytes()) {
+            throw new IllegalStateException("buffer queue length overflow: " + readableBytes
+                    + " + " + buf.readableBytes());
+        }
```

Comment:
Yes, HTTP2 requires supporting writing an empty DATA frame with end-of-stream=true. Plus you can have empty buffers with capacity > 0 that might need to be released.